contrastive_vi/module/total_contrastive_vi.py

```diff
@@ -0,0 +1,738 @@
+        n_samples: int = 1,
+    ) -> Dict[str, torch.Tensor]:
+        x_ = x
+        y_ = y
+        if self.use_observed_lib_size:
+            library_gene = x.sum(1).unsqueeze(1)
+        x_ = torch.log(1 + x_)
+        y_ = torch.log(1 + y_)
+        encoder_input = torch.cat((x_, y_), dim=-1)
+
+        (
+            qz_m,
+            qz_v,
+            ql_m,
+            ql_v,
+            background_latent,
+            untran_background_latent,
+        ) = self.z_encoder(encoder_input, batch_index)
+        z = background_latent["z"]
+        untran_z = untran_background_latent["z"]
+        untran_l = untran_background_latent["l"]  # Library encoder used and updated.
+        (qs_m, qs_v, _, _, salient_latent, untran_salient_latent) = self.s_encoder(
+            encoder_input, batch_index
+        )
+        s = salient_latent["z"]
+        untran_s = untran_salient_latent["z"]
+        # Library encoder not used and not updated.
+
+        if not self.use_observed_lib_size:
+            library_gene = background_latent["l"]
+
+        if n_samples > 1:
+            qz_m = self._reshape_tensor_for_samples(qz_m, n_samples)
+            qz_v = self._reshape_tensor_for_samples(qz_v, n_samples)
+            untran_z = Normal(qz_m, qz_v.sqrt()).sample()
+            z = self.z_encoder.z_transformation(untran_z)
+
+            qs_m = self._reshape_tensor_for_samples(qs_m, n_samples)
+            qs_v = self._reshape_tensor_for_samples(qs_v, n_samples)
+            untran_s = Normal(qs_m, qs_v.sqrt()).sample()
+            s = self.s_encoder.z_transformation(untran_s)
+
+            ql_m = self._reshape_tensor_for_samples(ql_m, n_samples)
+            ql_v = self._reshape_tensor_for_samples(ql_v, n_samples)
+            untran_l = Normal(ql_m, ql_v.sqrt()).sample()
+
+            if self.use_observed_lib_size:
+                library_gene = self._reshape_tensor_for_samples(library_gene, n_samples)
+            else:
+                library_gene = self.z_encoder.l_transformation(untran_l)
+
+        if self.n_batch > 0:
+            py_back_alpha_prior = F.linear(
+                one_hot(batch_index, self.n_batch), self.background_pro_alpha
+            )
+            py_back_beta_prior = F.linear(
+                one_hot(batch_index, self.n_batch),
+                torch.exp(self.background_pro_log_beta),
+            )
+        else:
+            py_back_alpha_prior = self.background_pro_alpha
+            py_back_beta_prior = torch.exp(self.background_pro_log_beta)
+
+        back_mean_prior = Normal(py_back_alpha_prior, py_back_beta_prior)
+
+        outputs = dict(
+            untran_z=untran_z,
+            z=z,
+            qz_m=qz_m,
+            qz_v=qz_v,
+            untran_s=untran_s,
+            s=s,
+            qs_m=qs_m,
+            qs_v=qs_v,
+            library_gene=library_gene,
+            ql_m=ql_m,
+            ql_v=ql_v,
+            untran_l=untran_l,
+            back_mean_prior=back_mean_prior
```

Comment:
Due to a subtle change in how scvi-tools thinks about batches, I had to add `back_mean_prior` here as an output so that it could be accessed later in our loss function. In short: while `self.n_batch` used to be 0 if all data came from a single batch, now it is one if you have a single batch. Thus, the code path starting at line 361 will always be taken, while previously we always were using the else statement (since we never used batch labels with this model).

As it turns out, our previous implementation logic broke with the new code path due to some tensor shape issues. Originally the `back_mean_prior` variable on line 373 was set as a class attribute (i.e., the line was `self.back_mean_prior = ...`). This was a problem because (1) we were concatenating the target + background samples before calling this function but (2) calling the `_generic_loss` function separately for target and background samples. Thus, the shape of `self.back_mean_prior` assumed that both target and background were being considered at the same time, while we were handling them individually in the loss function (--> shape errors occured). To fix this, I just added `back_mean_prior` to the output here and called `_generic_inference` separately for target and background samples in the `inference` function.

(I realize that was probably confusing so feel free to ping me with questions).